utilities.c

```diff
@@ -0,0 +1,213 @@
+        return NULL;
+    /* Application data */
+    memcpy(buffer, aligned_buffer + APPDATA_OFFSET, APPDATA_SIZE);
+    buffer[APPDATA_SIZE - 1] = '\0';
+
+    struct volume_info *result = malloc(sizeof(struct volume_info));
+    result->skipsectors = 20;
+    result->supported = 0;
+    result->fragmentcount = 20;
+    result->offset = offset;
+    result->isosize = isosize(aligned_buffer);
+
+    free(aligned_buffer);
+
+    for (size_t index = 0; index < APPDATA_SIZE;) {
+        size_t len;
+        if ((len = matches(buffer + index, "ISO MD5SUM = "))) {
+            index += len;
+            if (index + HASH_SIZE >= APPDATA_SIZE)
+                goto fail;
+
+            memcpy(result->hashsum, buffer + index, HASH_SIZE);
+            result->hashsum[HASH_SIZE] = '\0';
+            index += HASH_SIZE;
+            task |= TASK_MD5;
+            /* Skip to next semicolon. */
+            for (char *p = buffer + index; index < APPDATA_SIZE && *p != ';';
+                 p++, index++) {
+            }
+        } else if ((len = matches_number(buffer, index,
+                                         "SKIPSECTORS = ", (long int *) &result->skipsectors))) {
+            index = len;
+            if (index >= APPDATA_SIZE)
+                goto fail;
+            task |= TASK_SKIP;
+        } else if ((len = matches_number(buffer, index,
+                                         "RHLISOSTATUS=", (long int *) &result->supported))) {
+            index = len;
+            task |= TASK_SUPPORTED;
+        } else if ((len = matches(buffer + index, "FRAGMENT SUMS = "))) {
+            index += len;
+            if (index + FRAGMENT_SUM_SIZE >= APPDATA_SIZE)
+                goto fail;
+            memcpy(result->fragmentsums, buffer + index, FRAGMENT_SUM_SIZE);
+            result->fragmentsums[FRAGMENT_SUM_SIZE] = '\0';
+            task |= TASK_FRAGSUM;
+            index += FRAGMENT_SUM_SIZE;
+            for (char *p = buffer + index; index < APPDATA_SIZE && *p != ';';
+                 p++, index++) {
+            }
+        } else if ((len = matches_number(buffer, index, "FRAGMENT COUNT = ", (long int *) &result->fragmentcount))) {
+            index = len;
+            task |= TASK_FRAGCOUNT;
+        }
+        /* Either something is wrong or it skips a semicolon. */
+        index++;
+        if (task == TASK_DONE)
+            break;
+    }
+
+    if (task < TASK_SKIP + TASK_MD5) {
+    fail:
+        free(result);
+        return NULL;
+    }
+    return result;
+}
+
+bool validate_fragment(const MD5_CTX *const hashctx, const size_t fragment,
+                       const size_t fragmentsize, const char *const fragmentsums, char *hashsums) {
+    unsigned char digest[HASH_SIZE / 2];
+    MD5_CTX ctx;
+    memcpy(&ctx, hashctx, sizeof(ctx));
+    MD5_Final(digest, &ctx);
+    size_t j = (fragment - 1) * fragmentsize;
+    for (size_t i = 0; i < MIN(fragmentsize, HASH_SIZE / 2); i++) {
+        char tmp[2];
```

Comment:
I think the relevant part of the manpage is from the `Field Width` section: "In no case does a nonexistent or small field width cause truncation of a field;" so `%01` will not output 1, it will output 2 bytes. And gcc in a mockbuild complains about it like this:
```
utilities.c: In function 'validate_fragment':
utilities.c:194:26: error: '__builtin___snprintf_chk' output may be truncated before the last format character [-Werror=format-truncation=]
         snprintf(tmp, 2, "%01x", digest[i]);
                          ^~~~~~
In file included from /usr/include/stdio.h:889:0,
                 from utilities.c:19:
/usr/include/bits/stdio2.h:64:10: note: '__builtin___snprintf_chk' output between 2 and 3 bytes into a destination of size 2
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```